Stack/Opc.Ua.Core/Stack/Tcp/TcpListenerChannel.cs

```diff
@@ -0,0 +1,560 @@
+                buffers = null;
+            }
+            catch (Exception e)
+            {
+                if (buffers != null)
+                {
+                    buffers.Release(BufferManager, "SendServiceFault");
+                }
+
+                ForceChannelFault(ServiceResult.Create(e, StatusCodes.BadTcpInternalError, "Unexpected error sending a service fault."));
+            }
+        }
+
+        protected virtual void NotifyMonitors(ServiceResult status, bool closed)
+        {
+            // TODO: how is this used?
+        }
+
+        protected virtual void CompleteReverseHello(Exception e)
+        {
+            // TODO: if callback is only used in serverchannel, move implementation
+        }
+
+        /// <summary>
+        /// Sends a fault response secured with the asymmetric keys.
+        /// </summary>
+        protected void SendServiceFault(uint requestId, ServiceResult fault)
+        {
+            Utils.Trace("Channel {0} Request {1}: SendServiceFault()", ChannelId, requestId);
+
+            BufferCollection chunksToSend = null;
+
+            try
+            {
+                // construct fault.
+                ServiceFault response = new ServiceFault();
+
+                response.ResponseHeader.ServiceResult = fault.Code;
+
+                StringTable stringTable = new StringTable();
+
+                response.ResponseHeader.ServiceDiagnostics = new DiagnosticInfo(
+                    fault,
+                    DiagnosticsMasks.NoInnerStatus,
+                    true,
+                    stringTable);
+
+                response.ResponseHeader.StringTable = stringTable.ToArray();
+
+                // serialize fault.
+                byte[] buffer = BinaryEncoder.EncodeMessage(response, Quotas.MessageContext);
+
+                // secure message.
+                chunksToSend = WriteAsymmetricMessage(
+                    TcpMessageType.Open,
+                    requestId,
+                    ServerCertificate,
+                    ClientCertificate,
+                    new ArraySegment<byte>(buffer, 0, buffer.Length));
+
+                // write the message to the server.
+                BeginWriteMessage(chunksToSend, null);
+                chunksToSend = null;
+            }
+            catch (Exception e)
+            {
+                if (chunksToSend != null)
+                {
+                    chunksToSend.Release(BufferManager, "SendServiceFault");
+                }
+
+                ForceChannelFault(ServiceResult.Create(e, StatusCodes.BadTcpInternalError, "Unexpected error sending a service fault."));
+            }
+        }
+
+        /// <summary>
+        /// Handles a reconnect request.
+        /// </summary>
+        public virtual void Reconnect(IMessageSocket socket, uint requestId, uint sequenceNumber, X509Certificate2 clientCertificate, ChannelToken token, OpenSecureChannelRequest request)
+        {
+            throw new NotImplementedException();
```

Comment:
>throw new NotImplementedException(); [](start = 12, length = 36)

Implementation missing

Reply:
added to readme

---
In reply to: [432433528](https://github.com/OPCFoundation/UA-.NETStandard/pull/998#discussion_r432433528) [](ancestors = 432433528)